kv_store/lww_kvs/kvs_proxy.cpp

```diff
@@ -100,90 +163,344 @@ int main(int argc, char* argv[]) {
+        communication::Key_Response_Tuple* tp = res.add_tuple();
+        tp->set_key(map_it->first);
+        for (auto set_it = map_it->second.begin(); set_it != map_it->second.end(); set_it++) {
+          communication::Key_Response_Address* ad = tp->add_address();
+          ad->set_addr(*set_it);
+        }
+      }
+
+      // send the key address response to the server node
+      string response;
+      res.SerializeToString(&response);
+      zmq_util::send_string(response, &gossip_responder);
+    } else if (pollitems[3].revents & ZMQ_POLLIN) {
+      cerr << "received replication factor change request\n";
+      string placement_req = zmq_util::recv_string(&replication_factor_change_puller);
+      communication::Placement_Request req;
+      req.ParseFromString(placement_req);
+
+      // used to keep track of the original replication factors for the requested keys
+      unordered_map<string, pair<int, int>> orig_placement_info;
+
+      // used to keep track of the key value mapping
+      unordered_map<string, pair<string, int>> key_value_map;
+      unordered_map<address_t, communication::Request> request_map;
+      unordered_map<master_node_t, communication::Key_Request, node_hash> key_request_map;
+
+      for (int i = 0; i < req.tuple_size(); i++) {
+        string key = req.tuple(i).key();
+
+        vector<master_node_t> server_nodes = get_nodes(key, &global_memory_hash_ring, &global_ebs_hash_ring, placement);
+
+        // randomly choose a server node responsible for this key and update the key request map
+        // assume the size of server_nodes is not 0 (it shouldn't be) 
+        master_node_t server_node = server_nodes[rand() % server_nodes.size()];
+        // TODO: before setting the sender, check if it's already been set
+        key_request_map[server_node].set_sender("proxy");
+        communication::Key_Request_Tuple* tp = key_request_map[server_node].add_tuple();
+        tp->set_key(key);
+        tp->set_global_memory_replication(placement[key].global_memory_replication_);
+        tp->set_global_ebs_replication(placement[key].global_ebs_replication_);
+      }
+
+      // loop through key request map, send key request to all nodes
+      // receive key responses, and form the request map
+      for (auto it = key_request_map.begin(); it != key_request_map.end(); it++) {
+        // serialize request and send
+        string key_req;
+        it->second.SerializeToString(&key_req);
+        zmq_util::send_string(key_req, &requesters[it->first.key_exchange_connect_addr_]);
+
+        // wait for a response from the server and deserialize
+        string key_res = zmq_util::recv_string(&requesters[it->first.key_exchange_connect_addr_]);
+        communication::Key_Response server_res;
+        server_res.ParseFromString(key_res);
+
+        string key;
+        address_t worker_address;
+        // get the worker address from the response and sent the serialized
```

Comment:
Same point about repeated code. We should definitely think about factoring things out.

Reply:
yeah there are lots of code blocks that are very similar but have subtle difference...I'll try to do the refactoring tonight :)

Reply:
Addressed the refactoring comments

Reply:
There still seems to be repeated code here. L333-350 seems to be the same as here to 430.